src/gt4py/analysis/passes.py

```diff
@@ -1194,38 +1194,37 @@ class DemoteLocalTemporariesToVariablesPass(TransformPass):
     """
 
     class CollectDemotableSymbols(gt_ir.IRNodeVisitor):
-        def __init__(self):
-            pass
-
         @classmethod
         def apply(cls, node: gt_ir.StencilImplementation) -> Set[str]:
             collector = cls()
             return collector(node)
 
         def __call__(self, node: gt_ir.StencilImplementation) -> Set[str]:
             assert isinstance(node, gt_ir.StencilImplementation)
-            self.demotables = set(node.temporary_fields)
-            self.seen_symbols: Set[str] = set()
-            self.stage_symbols: Optional[Set[str]] = None
+            self.demotables: Dict[str, str] = {
+                field_name: "" for field_name in node.temporary_fields
+            }
             self.visit(node)
-            return self.demotables
+            return set(self.demotables.keys())
 
         def visit_Stage(self, node: gt_ir.Stage) -> None:
-            self.stage_symbols = set()
-
+            self.stage_name = node.name
             self.generic_visit(node)
```

Comment:
I think `self.generic_visit` can pass `stage_name` down to `visit_Assign` by adding it as a keyword to the signature of `visit_Assign`. In that case `stage_name` does not need to be a member of the class.
```
def visit_Assign(self, node: gt_ir.Assign, *, stage_name: str) -> None:
    ...
```